homeassistant/helpers/event.py

```diff
@@ -282,41 +294,395 @@ def remove_listener() -> None:
 
 track_template = threaded_listener_factory(async_track_template)
 
 
+_UNCHANGED = object()
+
+
+class _TrackTemplateResultInfo:
+    """Handle removal / refresh of tracker."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        template: Template,
+        action: Callable,
+        variables: Optional[TemplateVarsType],
+    ):
+        """Handle removal / refresh of tracker init."""
+        self.hass = hass
+        self._template = template
+        self._action = action
+        self._variables = variables
+        self._last_result: Optional[str] = None
+        self._last_exception = False
+        self._all_listener: Optional[Callable] = None
+        self._domains_listener: Optional[Callable] = None
+        self._entities_listener: Optional[Callable] = None
+        self._info: Optional[RenderInfo] = None
+        self._last_info: Optional[RenderInfo] = None
+
+    def async_setup(self) -> None:
+        """Activation of template tracking."""
+        self._info = self._template.async_render_to_info(self._variables)
+        if self._info.exception:
+            self._last_exception = True
+            _LOGGER.exception(self._info.exception)
+        self._create_listeners()
+        self._last_info = self._info
+
+    @property
+    def _needs_all_listener(self) -> bool:
+        assert self._info
+
+        # Tracking all states
+        if self._info.all_states:
+            return True
+
+        # Previous call had an exception
+        # so we do not know which states
+        # to track
+        if self._info.exception:
+            return True
+
+        # There are no entities in the template
+        # to track so this template will
+        # re-render on EVERY state change
+        if not self._info.domains and not self._info.entities:
+            return True
+
+        return False
+
+    @callback
+    def _create_listeners(self) -> None:
+        assert self._info
+
+        if self._info.is_static:
+            return
+
+        if self._needs_all_listener:
+            self._setup_all_listener()
+            return
+
+        if self._info.domains:
+            self._setup_domains_listener()
+
+        if self._info.entities or self._info.domains:
+            self._setup_entities_listener()
+
+    @callback
+    def _cancel_domains_listener(self) -> None:
+        if self._domains_listener is None:
+            return
+        self._domains_listener()
+        self._domains_listener = None
+
+    @callback
+    def _cancel_entities_listener(self) -> None:
+        if self._entities_listener is None:
+            return
+        self._entities_listener()
+        self._entities_listener = None
+
+    @callback
+    def _cancel_all_listener(self) -> None:
+        if self._all_listener is None:
+            return
+        self._all_listener()
+        self._all_listener = None
+
+    @callback
+    def _update_listeners(self) -> None:
+        assert self._info
+        assert self._last_info
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)